Add bulk ActivityPub capability removal confirmation

pfefferle opened the pull request:
Introduces a confirmation step for bulk removal of the ActivityPub capability from users in the admin panel. Adds a new template for confirmation, options to notify the Fediverse, and success notices after completion. Ensures safer and more transparent bulk user management.

<img width="1171" height="1063" alt="Screenshot 2025-09-08 at 18 13 57" src="https://github.com/user-attachments/assets/dd4fccb4-ceec-414f-ad57-2986a2fd557a" />

## Proposed changes:
- Introduces confirmation page template with individual user control for Fediverse notifications
- Updates bulk action handler to redirect to confirmation page instead of immediate removal
- Adds unified capability removal processing with success notifications

### Other information:

- [ ] Have you written new tests for your changes, if applicable?

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Go to the Users page and bulk-remove the ActivityPub capability from selected users.
* On the approval page, test different settings:
  * Select some users (checkbox) to confirm they will be added as Delete activities in the outbox.
  * Cancel the request and confirm that the capability is removed but no Delete activities are added.

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger a GitHub workflow that will create and push the entry into the branch. -->

<!-- Due to org permissions, the job may fail for PRs crated from a fork under GitHub organizations. -->
<!-- In this case, you can create entry manually with `composer changelog:add` and push it into the branch. -->

<!-- If no changelog entry is required for this PR, please add the "Skip Changelog" label. -->

-   [X] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [X] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [X] Added - for new features
-   [ ] Changed - for changes in existing functionality
-   [ ] Deprecated - for soon-to-be removed features
-   [ ] Removed - for now removed features
-   [ ] Fixed - for any bug fixes
-   [ ] Security - in case of vulnerabilities

#### Message

Added confirmation step for bulk removal of ActivityPub capability, asking whether to also delete users from the Fediverse.

</details>

pfefferle committed:
Add bulk ActivityPub capability removal confirmation

Introduces a confirmation step for bulk removal of the ActivityPub capability from users in the admin panel. Adds a new template for confirmation, options to notify the Fediverse, and success notices after completion. Ensures safer and more transparent bulk user management.

pfefferle committed:
Refactor bulk ActivityPub capability removal flow

Simplifies the bulk capability removal confirmation template to use per-user checkboxes for Fediverse removal, updates admin logic to process these selections, and improves debug logging. Also updates scheduling logic to allow bypassing capability checks when needed.

pfefferle committed:
Refactor bulk capability removal and user delete logic

Streamlines bulk removal of ActivityPub capability and user deletion by improving code clarity, removing debug logging, and ensuring delete activities are scheduled before capability removal. Removes bypass_cap_check from Actor::schedule_user_delete and updates related admin logic and template variable formatting for consistency.

pfefferle committed:
Refactor bulk ActivityPub capability removal logic

Unified the capability removal and fediverse deletion logic into a single method, process_capability_removal, simplifying bulk user operations and reducing code duplication. Legacy and new formats for fediverse deletion are now supported, and cache refreshes are consistently applied after capability changes.

pfefferle committed:
Sanitize input data in bulk capability removal

Added sanitize_text_field to all GET and POST parameters in class-admin.php to improve security and data integrity. Minor code style improvements were made in the bulk-capability-removal-confirmation.php template for consistency.

pfefferle committed:
Remove bulk capability removal success notices

Eliminated the display of success notices for bulk ActivityPub capability removal in the admin interface. Also added phpcs ignore comments for nonce verification checks in the bulk capability removal handler.

pfefferle committed:
Merge branch 'trunk' into change/delete-user-from-fediverse-by-capability

pfefferle committed:
Remove sanitize_text_field from array inputs

Sanitization with sanitize_text_field was removed from array inputs for 'users', 'selected_users', and 'remove_from_fediverse' to avoid improper handling of array data. PHPCS ignore comments were updated to reflect the change.

matticbot committed:
Add changelog

pfefferle committed:
Refactor ActivityPub capability removal flow

Capabilities are now removed immediately before fediverse deletion confirmation. The confirmation template and related logic have been updated to reflect this change, clarifying that fediverse deletion is optional and occurs after capability removal.

pfefferle committed:
Remove redundant comment and whitespace in bulk capability removal

Deleted an unnecessary comment in the bulk capability removal confirmation template and removed extraneous whitespace in the admin class. These changes improve code clarity without affecting functionality.

obenland commented on includes/wp-admin/class-admin.php:
```diff
@@ -593,23 +598,207 @@ public static function user_bulk_options( $actions ) {
 			'remove_activitypub_cap' !== $action &&
-			'add_activitypub_cap' !== $action
+			'add_activitypub_cap' !== $action &&
+			'remove_activitypub_cap_confirmed' !== $action
```
With these updates we could probably simplify this function to have a single switch statement and not worry about returning early.

obenland commented on includes/wp-admin/class-admin.php:
```diff
@@ -593,23 +598,207 @@ public static function user_bulk_options( $actions ) {
+			// Build the query args with proper array handling for fediverse deletion confirmation.
+			$query_args = array(
+				'activitypub_confirm_removal' => '1',
+				'send_back'                   => rawurlencode( $send_back ),
```
```suggestion
				'send_back'                   => \rawurlencode( $send_back ),
```

obenland commented on includes/wp-admin/class-admin.php:
```diff
@@ -593,23 +598,207 @@ public static function user_bulk_options( $actions ) {
+		}
+
+		// This should be handled by handle_bulk_capability_removal_confirmation
+		// but we include it here as a fallback.
```
```suggestion
		// This should be handled by handle_bulk_capability_removal_confirmation but we include it here as a fallback.
```
Minor: Let's keep it on a single line or use [multiline comment docs](https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#5-2-multi-line-comments).

obenland commented on includes/wp-admin/class-admin.php:
```diff
@@ -593,23 +598,207 @@ public static function user_bulk_options( $actions ) {
+
+			// Build the query args with proper array handling for fediverse deletion confirmation.
+			$query_args = array(
+				'activitypub_confirm_removal' => '1',
```
```suggestion
				'action' => 'activitypub_confirm_removal',
```
With this change, could we move `handle_bulk_capability_removal_page()` to the `admin_action_activitypub_confirm_removal` action (instead of `current_screen`)?

That way `handle_bulk_capability_removal_page()` wouldn't have to check for `is_admin()` or whether `activitypub_confirm_removal` is set (and it doesn't execute on every admin page load).

obenland commented on includes/wp-admin/class-admin.php:
```diff
@@ -593,23 +598,207 @@ public static function user_bulk_options( $actions ) {
+			$message_params['activitypub_bulk_deleted'] = $deleted_count;
+		}
+
+		return \add_query_arg( $message_params, $send_back );
```
```suggestion
		if ( $deleted_count ) {
			$send_back = \add_query_arg( 'activitypub_bulk_deleted', $deleted_count, $send_back );
		}

		return $send_back;
```
Simplifies it a bit.

It doesn't look like we do anything with `activitypub_bulk_deleted`? I didn't get an admin notice after removing caps from multiple users.

obenland commented on includes/wp-admin/class-admin.php:
```diff
@@ -593,23 +598,207 @@ public static function user_bulk_options( $actions ) {
+			\add_filter( 'activitypub_user_can_activitypub', '__return_true' );
+
+			foreach ( $users as $user_id ) {
+				if ( isset( $users_to_delete[ $user_id ] ) ) {
```
I'm not sure why we need to flip `$users_to_delete` and check for overlap with `$users`. What keeps us from just doing `array_map( array( Actor::class, 'schedule_user_delete' ), $users_to_delete );` and use `count( $users_to_delete )` as the `$deleted_count`?

obenland commented on templates/bulk-capability-removal-confirmation.php:
```diff
@@ -0,0 +1,94 @@
+	<?php
+	printf(
+		/* translators: %d: number of users */
+		esc_html( _n( 'ActivityPub capability has been removed from %d user. Do you want to also remove them from the Fediverse?', 'ActivityPub capability has been removed from %d users. Do you want to also remove them from the Fediverse?', $user_count, 'activitypub' ) ),
```
I think it would be helpful for users to understand why sending out delete activities is an optional step. Reading this for the first time I'd wonder "Why _wouldn't_ I want to remove them from the Fediverse"?

The sentence in line 85 might be more helpful up here, with a brief summary of what that'll do. Or why they wouldn't want to send it, like when it's a temporary revocation.

obenland commented on templates/bulk-capability-removal-confirmation.php:
```diff
@@ -0,0 +1,94 @@
+
+		<p class="submit">
+			<?php submit_button( __( 'Delete from Fediverse', 'activitypub' ), 'primary', 'submit', false ); ?>
+			<a href="<?php echo esc_url( $send_back ); ?>" class="button"><?php esc_html_e( 'Skip Fediverse Deletion', 'activitypub' ); ?></a>
```
```suggestion
			<a href="<?php echo esc_url( $send_back ); ?>" class="button"><?php esc_html_e( 'Skip', 'activitypub' ); ?></a>
```
I think the context is clear, there's an opportunity to be more concise

obenland commented on templates/bulk-capability-removal-confirmation.php:
```diff
@@ -0,0 +1,94 @@
+		'display_name' => $user->display_name,
+		'email'        => $user->user_email,
+	);
+}
```
Not sure if `get_users( array( 'include' => $users ) );` and then just using field names directly might be good enough here?

pfefferle committed:
Update includes/wp-admin/class-admin.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Update includes/wp-admin/class-admin.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Update includes/wp-admin/class-admin.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Update includes/wp-admin/class-admin.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Update templates/bulk-capability-removal-confirmation.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Refactor bulk request handler with switch statement

Replaces multiple if statements with a switch for handling bulk user capability actions, improving readability and maintainability. Logic for adding and removing ActivityPub capability remains unchanged.

pfefferle committed:
Update bulk capability removal page check

Refines the confirmation page request logic to check for the 'action' parameter and its value instead of only 'activitypub_confirm_removal'. This improves clarity and robustness in request validation.

pfefferle committed:
Rename bulk capability removal to actor delete

Refactored method and action names in Admin class from 'bulk capability removal' to 'bulk actor delete' for clarity. Also renamed the confirmation template file to match the new terminology.

pfefferle committed:
Rename bulk actor deletion action and update confirmation UI

Changed the action and case name from 'remove_activitypub_cap_confirmed' to 'delete_actor_confirmed' for clarity and consistency. Updated the bulk actor deletion confirmation template to improve messaging, clarify the optional nature of Fediverse deletion, and add a note that users are not deleted from WordPress.

pfefferle committed:
Refactor user deletion logic for fediverse removal

Simplifies the process of determining users to delete from the fediverse by removing array flipping and using direct arrays. Replaces the user deletion loop with array_map for scheduling deletions and updates the success message to use the count of deleted users.

pfefferle committed:
Update bulk-actor-delete-confirmation.php

pfefferle committed:
Refactor bulk actor delete confirmation user handling

Simplifies user data preparation in the bulk actor delete confirmation template by using get_users directly and updating references accordingly. Removes redundant code and streamlines user selection and display logic. Also removes success message parameter addition from the admin redirect URL.

pfefferle committed:
Update action hook for actor delete page

Replaces the 'current_screen' action hook with 'admin_action_activitypub_confirm_removal' for handling bulk actor delete page logic in Admin class.

pfefferle commented:
@obenland thanks for your detailed feedback, I think I have addressed everything.

pfefferle committed:
Remove unused variable in capability removal method

Deleted the unused $deleted_count variable from the process_capability_removal method in class-admin.php to clean up the code.

pfefferle committed:
Update bulk actor deletion confirmation text

Revised messaging in the bulk-actor-delete-confirmation.php template to clarify the consequences of Fediverse deletion and improve user guidance. Added a new explanatory paragraph and adjusted wording for better clarity.

pfefferle committed:
Merge branch 'trunk' into change/delete-user-from-fediverse-by-capability

obenland commented on includes/wp-admin/class-admin.php:
```diff
@@ -591,20 +596,181 @@ public static function user_bulk_options( $actions ) {
+	 * Handle the bulk capability removal page request directly.
+	 */
+	public static function handle_bulk_actor_delete_page() {
+		// Check if this is our confirmation page request.
+		// phpcs:ignore WordPress.Security.NonceVerification
+		if ( ! isset( $_GET['action'] ) || 'activitypub_confirm_removal' !== $_GET['action'] ) {
+			return;
+		}
+
+		// Check if we're in admin area.
+		if ( ! \is_admin() ) {
+			return;
+		}
```
```suggestion
```

With it hooked to the `admin_action_*` hook, I think this can be removed

pfefferle committed:
Merge branch 'trunk' into change/delete-user-from-fediverse-by-capability

obenland commented on templates/bulk-actor-delete-confirmation.php:
```diff
@@ -0,0 +1,65 @@
+ * @package Activitypub
+ */
+
+use Activitypub\Collection\Actors;
+
```
```suggestion
```

obenland commented on templates/bulk-actor-delete-confirmation.php:
```diff
@@ -0,0 +1,65 @@
+}
+
+// Prepare user data for display.
+$users = \get_users( array( 'include' => $users ) );
```
```suggestion
$users = get_users( array( 'include' => $users ) );
```

obenland commented on templates/bulk-actor-delete-confirmation.php:
```diff
@@ -0,0 +1,65 @@
+	exit;
+}
+
+require_once ABSPATH . 'wp-admin/admin-header.php';
```
```suggestion
$GLOBALS['plugin_page'] = '';
require_once ABSPATH . 'wp-admin/admin-header.php';
```

Fixes a deprecated notice. It might be worth investigating what's different in the load order between this template and the block confirmation template to understand why this deprecated notice shows up and whether this should be loaded from somewhere else.

```
Deprecated: preg_replace(): Passing null to parameter #3 ($subject) of type 
  array|string is deprecated in /var/www/html/wp-admin/includes/plugin.php:2161

  Stack Trace
  1.     preg_replace('!\.php!', '', NULL)
  /var/www/html/wp-admin/includes/plugin.php:2161
  2.     get_plugin_page_hookname(NULL, 'users.php')
  /var/www/html/wp-admin/includes/plugin.php:2125
  3.     get_plugin_page_hook(NULL, 'users.php')
  /var/www/html/wp-admin/includes/plugin.php:2049
  4.     get_admin_page_title()
  /var/www/html/wp-admin/admin-header.php:40
  5.     require_once('/var/www/html/wp-admin/admin-header.php')
  /var/www/html/wp-content/plugins/wordpress-activitypub/templates/bulk-actor-delete-confirmation.php:30
  7.     load_template('/var/www/html/wp-content/plugins/wordpress-activitypub/templates/bulk-actor-delete-confirmation.php
  ', false, Array)
  /var/www/html/wp-content/plugins/wordpress-activitypub/includes/wp-admin/class-admin.php:688
  8.     Activitypub\WP_Admin\Admin::handle_bulk_actor_delete_page('')
  /var/www/html/wp-includes/class-wp-hook.php:324
  9.     WP_Hook::apply_filters('', Array)
  /var/www/html/wp-includes/class-wp-hook.php:348
  10.     WP_Hook::do_action(Array)
  /var/www/html/wp-includes/plugin.php:517
  12.     require_once('/var/www/html/wp-admin/admin.php')
  /var/www/html/wp-admin/users.php:11
```

obenland commented on templates/bulk-actor-delete-confirmation.php:
```diff
@@ -0,0 +1,65 @@
+<div class="wrap">
+	<h1><?php esc_html_e( 'Delete Users from Fediverse', 'activitypub' ); ?></h1>
+	<p><?php esc_html_e( 'You&#8217;ve just removed the capability to publish to the Fediverse for the selected users, do you want to also remove them from the Fediverse?', 'activitypub' ); ?></p>
+	<p><?php echo wp_kses( __( 'Fediverse deletion is optional but recommended to properly notify your followers, <strong>but be aware that this action might be irreversible.</strong>', 'activitypub' ), array( 'strong' => array() ) ); ?></p>
```
```suggestion
	<p><?php echo wp_kses( __( 'Fediverse deletion is optional but recommended to properly notify your followers. <strong>This action is irreversible.</strong>', 'activitypub' ), array( 'strong' => array() ) ); ?></p>
```
There are two "but"s in this sentence, let's simplify it.

obenland commented on includes/wp-admin/class-admin.php:
```diff
@@ -591,20 +596,181 @@ public static function user_bulk_options( $actions ) {
+
+		// Sanitize user IDs.
+		$users = \array_map( 'absint', (array) $users );
+		$users = \array_filter( $users );
```
We should remove users that don't currently have the `activitypub` capability. I was able to get to the confirmation page by selecting any user and running the "Disable for Activitypub" bulk action.

pfefferle committed:
Update includes/wp-admin/class-admin.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Update templates/bulk-actor-delete-confirmation.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Update templates/bulk-actor-delete-confirmation.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Update templates/bulk-actor-delete-confirmation.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Update templates/bulk-actor-delete-confirmation.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Merge branch 'trunk' into change/delete-user-from-fediverse-by-capability

pfefferle committed:
Ignore PHPCS warning for global variable override

Added a phpcs:ignore comment to suppress the WordPress.WP.GlobalVariablesOverride.Prohibited warning for the $GLOBALS['plugin_page'] assignment in bulk-actor-delete-confirmation.php.

pfefferle committed:
Merge branch 'trunk' into change/delete-user-from-fediverse-by-capability

pfefferle committed:
Fix user removal logic in capability check

Updates the user removal loop to unset users from the array if they do not have the 'activitypub' capability, ensuring only eligible users are processed.

pfefferle commented:
@obenland should be ready to merge now.

pfefferle committed:
Merge branch 'trunk' into change/delete-user-from-fediverse-by-capability

obenland commented on templates/bulk-actor-delete-confirmation.php:
```diff
@@ -0,0 +1,64 @@
+			<a href="<?php echo esc_url( $send_back ); ?>" class="button"><?php esc_html_e( 'Skip', 'activitypub' ); ?></a>
+		</p>
+
+		<p><?php esc_html_e( '(The users will not be deleted from your WordPress installation.)', 'activitypub' ); ?></p>
```
Could we avoid having to add this if we rephrased the title to "Remove Users from Fediverse"? And the submit button I guess.

pfefferle replied:
Totally!

pfefferle committed:
Remove redundant user deletion notice

Deleted the message clarifying that users will not be deleted from WordPress, as it was unnecessary in the bulk actor delete confirmation template.

pfefferle committed:
Merge branch 'trunk' into change/delete-user-from-fediverse-by-capability
